### ISSUE TITLE
Design: diff page

### ISSUE THREAD

**Issue by sbcgua** [larshp/abapGit#213]
Branch from #202 (due to concurent coding of the main part ;)

New disign of DIFF

![image](https://cloud.githubusercontent.com/assets/15635498/15267905/2eb44ee6-19d7-11e6-879e-8b625bdc1b56.png)

Have a look [here](https://github.com/sbcgua/abapGit/tree/diffdesign). If OK, I'll create a merge request (after a few more tweaks yet)

1) html helper class is evolving (don't judge it critically) but the idea is to create html related routines to improve readability of the main code. I have already what to add and what to refactor there.
2) html helper variable naming as arguable - I'd like to keep them as short as possible. Initially they were even called like "h".
3) I didn't get why did you do anchors ... could you please explain ?
4) the line numberring should be probably base on actual code in local and remote
5) The logic of highlighting seems to be incorrect currently - it doesn't show an option when remote is change and local is not ... a bug ? ... or I don't understand something ...
6) "Back" button - I've asked a colleague of mine to have a look, she said it is possible to make it it via sap button - the code is coming soon.

**Comment by larshp** [comment]
looks cool

3: anchors, to jump to next difference, if the source file is 20000 lines, and first difference is at line 10000, it helped a lot being able to just click a link and it would automatically jump to line 10000

In the previous version it had a lot of obsolete stuff, this is removed in the latest version

4: yes, line numbers are needed for both local and remote. Actually added line numbers in version 1.6.0, just an hour ago, see https://github.com/larshp/abapGit/commit/7572a7ddeea39e1321185979aef0fa7d873d28ab

5: can you help trying to make an example, perhaps raise it as a separate issue

I like the idea with the yellow colour, in the current implementation it just shows both remote and local in green colour, having it as yellow will make it easier to understand :smile:

**Comment by larshp** [comment]
1: perhaps just submit this as a separate small PR

2: ok, just make sure to prefix the variable names, see https://github.com/larshp/abapGit/blob/master/CONTRIBUTING.md c

6: the users just need to be able to copy paste the report into the system for installation, no additional steps should be required. So a solution with adding a dynpro screen will not work, if this is the approach

**Comment by sbcgua** [comment]
2) names changed
3) done but differently (right line corner)
4) Nice ! Integrated
5) will do but later


7) Completely confused with this new piece if code with shortlist. What does it do ? I rewrote it in the new way (keeping the logic) but I really didn't understand the idea.

**Comment by sbcgua** [comment]
![image](https://cloud.githubusercontent.com/assets/15635498/15269384/a919f16c-1a05-11e6-979b-7b5472c5d57f.png)
screenshot

**Comment by larshp** [comment]
7: shortlist: earlier it showed the full local file and the full remote file. For eg. abapgit this means that it shows 20000+ lines in the GUI, this influence performance due to the amount of data being transferred from application server to SAP GUI, and rendering time in the browser.

Typically the user only need information regarding the code sourrounding the changed lines, just like github does, see eg. https://github.com/larshp/abapGit/pull/214/files

The shortlist will trigger if the file which diff is performed for is larger than 5000 lines, in this case it will only show partial remote and local file. It will show at least the 20 lines before the changed code.

**Comment by larshp** [comment]
hi,
I added a few unit tests regarding the automatic indentation, see https://github.com/larshp/abapGit/commit/ee1bf174a42c8efbff9317e0c1cf140b1746478b#diff-35f1b917526b3abedbc0c9478bc2b08fR19450

I have some bad habits from the 90'ies where I tend to use BR and INPUT tags without closing them. It is not proper XHTML, but perhaps we should consider just ignoring BR and INPUT tags when calculating the indentation?

**Comment by sbcgua** [comment]
Unit test: done

7) Short list: mmmm currently looks incomplete. would look better if showed with also 20 lines after. Generally I'd say that it (shotlist) should be default behavior for any code (>100 lines?). Blanks between sections can be used for commands potentetially, 5th column with link is not needed then - better UX. Potentially a command needed to switch to "long list".

**Comment by larshp** [comment]
7: yes, I've also been thinking of showing some code after the change, it is handled at
https://github.com/larshp/abapGit/blob/master/zabapgit.prog.abap#L1911

100 lines, ok

long list, yeah, I was just lazy when implementing the short list

feel free to change it :smile:

**Comment by larshp** [comment]
@sbcgua I've added you as a collaborator to the project :smile: for minor changes feel free to push directly to master

**Comment by sbcgua** [comment]
Thanks ! :)
P.S. And collaborator is my colleague who know how to get rid of "Back" links :) I'll commit changes in a moment

**Comment by sbcgua** [comment]
Adjusted diff short list, now it looks forward as well, lines limit put down to 500.
In UI there is now a nav line. Still some space for improvement. In particular, would be cool to have context output like in github ("CLASS xxxx ..."):

![image](https://cloud.githubusercontent.com/assets/15635498/15453627/825cd5bc-201d-11e6-94c1-3f5fcc62a411.png)

@larshp do you think it is difficult to implement ? Ideally without code parsing ? Does that `COMPUTE_DELTAS` provide this info ? If not what kind of context can/should be traced ? My guess: CLASS, FORM, MODULE. Anything else ? 

P.S. That's probably all commit for today ... need to do some other stuff. Nice reviewing 10 commits :)))

**Comment by larshp** [comment]
good idea, but it will require some work. It think we should close this issue and create a new with the above idea. So we can get rid of the quite long story in this issue
